### DISCUSSION TITLE
Clarifying some misunderstood points

### DISCUSSION THREAD

**Discussion by aperullo** [ory/hydra#3155] [Q&A] [answered]
Hi! Ory looks like an awesome set of projects with good docs and I wanted to play with it in my personal lab and maybe finally wrap my head around Oauth. None of it has auth yet which is what I'm looking to add. I'm not entirely sure which pieces I need though and I'm hoping to be pointed in the right direction.

If I understand correctly I want kratos as a SSO provider and Hydra for oauth? 

There are several Off-the-shelf apps that support oidc but also two custom things I want to secure, which is mainly what I think I'm confused on. Theres

- an API that commits to a git repo on the user's behalf. It'd be cool for those commits to have the user's name, so I need to be able to get that info from something (the bearer token? Kratos?)

- A cli which talks to that api. The cli stores the user's credentials. I think this would be an auth code flow? Or implicit?

The way I understand it:

1.  The cli should reach out to the API
2. which will redirect it to Hydra with some kind of state so the cli doesn't need the clientid or secret. 
3. Hydra redirects to Kratos to sign in... Maybe?
4. Shouldn't there be a redirect to a locally running server the cli started to get the token?
5. ...?
6. The cli gets a token it can use to auth with the api and the api can tell who it is. 

Is the cli or API the client? Do I need Kratos and Hydra for this or hydra and any sign in provider would do?

I would appreciate if anyone can help fill the holes in my understanding. Thank you

**Comment by vinckr** [comment, marked as answer]
Hello @aperullo 
There are probably multiple options how to solve this using Ory. 
Check out this [blogpost to protect APIs with Kong&Kratos](https://www.ory.sh/zero-trust-api-security-ory-tutorial/). 

for usecases such as 
>API that commits on the users behalf 
>A cli which talks to that api

The OAuth2 [client credentials flow with jwks](https://www.ory.sh/docs/hydra/advanced#oauth-20-client-credentials-grant) is what you want for the CLI authentication - so using Hydra to serve the OAuth2 flows / tokens and Kratos to take care of the authentication. 


See also these related discussions:
- [CLI Authentication](https://github.com/ory/kratos/discussions/1139)
- [API Keys](https://github.com/ory/kratos/discussions/1304)
- [API Authentication](https://github.com/ory/kratos/issues/385)

I hope that makes it a bit clearer. 
Would be great to have an example for API/CLI auth in [ory/examples](https://github.com/ory/examples) down the line 🤔

**Reply by vinckr** [reply]
Hey @aperullo ,
Just checking in if your question was answered or if there was still something unclear.
I will mark this as answered for now, but feel free to come back to the discussion any time.

Thanks a ton! 🐝
